modules/puppetca/puppetca_csr.rb

```diff
@@ -0,0 +1,25 @@
+module Proxy::PuppetCa
+  class CertificateRequest
+    attr_reader :csr
+
+    def initialize(raw_csr)
+      @csr = OpenSSL::X509::Request.new(raw_csr)
+    end
+
+    def challenge_password
+      attribute = custom_attributes.detect do |attr|
+        ['challengePassword', '1.2.840.113549.1.9.7'].include?(attr[:oid])
+      end
+      attribute ? attribute[:value] : nil
+    end
+
+    def custom_attributes
+      @csr.attributes.map do |attr|
+        {
+          oid: attr.oid,
+          value: attr.value.value.first.value
```

Comment:
This looks fragile: are all of the intermediate calls guaranteed to not return nils?

Reply:
I think this should be fine. Openssl is not known for its nice code.

Reply:
Yeah, would probably be better to catch possible errors from this.